File: app/CDoc/c/src/data/Dpath.c
// Copyright 08-Dic-2021 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "data/Dpath.h"
#include "dmc/DEFS.h"
#include "dmc/js.h"
#include "dmc/char/Achar.h"

Dpath *dpath_new (
  char *id,
  char *path,
  int is_shown,
  int is_valid
) {
  Dpath *this = MALLOC(Dpath);
  this->id = id;
  this->path = path;
  this->is_shown = is_shown;
  this->is_valid = is_valid;
  return this;
}

void dpath_set_shown (Dpath *this, int value) {
  this->is_shown = value;
}

void dpath_set_valid (Dpath *this, int value) {
  this->is_valid = value;
}

char *dpath_to_js (Dpath *this) {
  return js_wa(achar_new_from(
    js_ws(this->id),
    js_ws(this->path),
    js_wb(this->is_shown),
    js_wb(this->is_valid),
    NULL
  ));
}

Dpath *dpath_from_js (char *js) {
  char **p = js_ra(js)->es;
  Dpath *this = MALLOC(Dpath);
  this->id = js_rs(*p++);
  this->path = js_rs(*p++);
  this->is_shown = js_rb(*p++);
  this->is_valid = js_rb(*p++);
  return this;
}

File: app/CDoc/js/src/data/Dpath.js
// Copyright 18-Aug-2019 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

/**
    Documentation path data.
**/
export default class Dpath {

  /* .
  _rc_ Dpath: serial
    id: string
    path: string
    show: boolean
    valid: boolean
  */

  /*--*/
  /**
      @param {string} id
      @param {string} path
      @param {boolean} show
      @param {boolean} valid
  **/
  constructor (id, path, show, valid) {

    /**
        @private
        @type {string}
    **/
    this._id = id;

    /**
        @private
        @type {string}
    **/
    this._path = path;

    /**
        @private
        @type {boolean}
    **/
    this._show = show;

    /**
        @private
        @type {boolean}
    **/
    this._valid = valid;

  }

  /**
      @return {string}
  **/
  get id () {
    return this._id;
  }

  /**
      @return {string}
  **/
  get path () {
    return this._path;
  }

  /**
      @return {boolean}
  **/
  get show () {
    return this._show;
  }

  /**
      @return {boolean}
  **/
  get valid () {
    return this._valid;
  }

  /**
      @return {!Array<?>}
  **/
  toJs () {
    return [
      this._id,
      this._path,
      this._show,
      this._valid
    ];
  }

  /**
      @param {!Array<?>} serial
      @return {!Dpath}
  **/
  static fromJs (serial) {
    return new Dpath(
      serial[0],
      serial[1],
      serial[2],
      serial[3]
    );
  }
  /*--*/
}
